fathom/src/driver/compiler.rs

```diff
@@ -0,0 +1,164 @@
+            core_module: None,
+        }
+    }
+
+    pub fn parse_surface(&mut self, files: &SimpleFiles<String, String>) -> lang::surface::Module {
+        let file_id = self.input_fathom_file_id;
+        let source = files.source(file_id).unwrap();
+        let surface_module =
+            lang::surface::Module::parse(file_id, source, &mut self.found_messages);
+        self.surface_module = Some(surface_module.clone());
+
+        return surface_module;
+    }
+
+    pub fn elaborate(
+        &mut self,
+        files: &SimpleFiles<String, String>,
+        surface_module: &lang::surface::Module,
+    ) -> lang::core::Module {
+        let mut context = surface_to_core::Context::new(&GLOBALS);
+        let core_module = context.from_module(&surface_module);
+        self.found_messages.extend(context.drain_messages());
+
+        // The core syntax from the elaborator should always be well-formed!
+        // store context in Compiler object?
+        let mut context = lang::core::typing::Context::new(&GLOBALS);
+        context.is_module(&core_module);
+        let validation_messages = context.drain_messages().collect::<Vec<_>>();
+
+        if !validation_messages.is_empty() {
+            let pretty_arena = pretty::Arena::new();
+            let mut buffer = BufferWriter::stderr(ColorChoice::Auto).buffer();
+
+            for message in &validation_messages {
+                let diagnostic = message.to_diagnostic(&pretty_arena);
+                term::emit(&mut buffer, &self.term_config, files, &diagnostic).unwrap();
+            }
+
+            eprintln!("  • elaborate: validate");
+            eprintln!();
+            eprintln_indented(4, "", "---- found diagnostics ----");
+            eprintln_indented(4, "| ", &String::from_utf8_lossy(buffer.as_slice()));
+            eprintln!();
+        }
+        self.core_module = Some(core_module.clone());
+
+        core_module
+    }
+
+    pub fn finish(mut self, files: &SimpleFiles<String, String>) {
+        let pretty_arena = pretty::Arena::new();
+        let found_diagnostics: Vec<Diagnostic<usize>> = self
+            .found_messages
+            .iter()
+            .map(|message| message.to_diagnostic(&pretty_arena))
+            .collect();
+
+        if !found_diagnostics.is_empty() {
+            self.failed_checks.push("unexpected_diagnostics");
+
+            eprintln!("Unexpected errors found:");
+            eprintln!();
+
+            // Use a buffer so that this doesn't get printed interleaved with the
+            // test status output.
+
+            let mut buffer = BufferWriter::stderr(ColorChoice::Auto).buffer();
+            for diagnostic in &found_diagnostics {
+                term::emit(&mut buffer, &self.term_config, files, diagnostic).unwrap();
+            }
+
+            eprintln_indented(4, "", "---- found diagnostics ----");
+            eprintln_indented(4, "| ", &String::from_utf8_lossy(buffer.as_slice()));
+            eprintln!();
```

Comment:
Probably don't need this kind of output for the CLI - perhaps the client of the compiler could handle the diagnostic rendering instead? So there's a `.drain_messages` method on the compiler, perhaps.

Reply:
Stuff like `---- found diagnostics ----` was there for the tests to distinguish it from the regular rust/test harness output.

Reply:
Fair enough!  Just so i understand, are you suggesting that the compiler simply expose the list of messages, and the CLI handle their formatting?